File: swimiq_app/lib/screens/dashboard_screen.dart
import 'package:fl_chart/fl_chart.dart';
import 'package:flutter/material.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../config/theme.dart';
import '../providers/app_providers.dart';
import '../utils/personal_bests.dart';
import '../utils/swim_time.dart';
import '../utils/swimiq_score.dart';
import '../widgets/empty_state.dart';
import '../widgets/metric_card.dart';
import '../widgets/section_header.dart';

class DashboardScreen extends ConsumerWidget {
  const DashboardScreen({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
    final dataAsync = ref.watch(swimmerDataProvider);

    return dataAsync.when(
      loading: () => const Center(child: CircularProgressIndicator()),
      error: (error, _) => Center(child: Text('Could not load dashboard: $error')),
      data: (data) {
        if (data.raceLogs.isEmpty) {
          return const EmptyState(
            icon: Icons.dashboard_outlined,
            title: 'No swim sessions yet',
            message: 'Add a swim session to start building your dashboard.',
          );
        }

        final score = SwimIQScore.calculate(data.raceLogs, data.goals);
        final pbs = PersonalBests.fromRaceLogs(data.raceLogs);
        final times = data.raceLogs.map((log) => log.timeSeconds).toList();
        final bestTime = times.reduce((a, b) => a < b ? a : b);
        final avgTime = times.reduce((a, b) => a + b) / times.length;

        return RefreshIndicator(
          onRefresh: () async => refreshData(ref),
          child: ListView(
            padding: const EdgeInsets.all(16),
            children: [
              GridView.count(
                crossAxisCount: 2,
                shrinkWrap: true,
                physics: const NeverScrollableScrollPhysics(),
                mainAxisSpacing: 12,
                crossAxisSpacing: 12,
                childAspectRatio: 1.5,
                children: [
                  MetricCard(
                    label: 'SwimIQ Score',
                    value: '$score',
                    icon: Icons.star_rounded,
                    highlight: true,
                  ),
                  MetricCard(
                    label: 'Total Sessions',
                    value: '${data.raceLogs.length}',
                    icon: Icons.fitness_center_rounded,
                  ),
                  MetricCard(
                    label: 'Personal Bests',
                    value: '${pbs.length}',
                    icon: Icons.emoji_events_outlined,
                  ),
                  MetricCard(
                    label: 'Active Goals',
                    value: '${data.goals.length}',
                    icon: Icons.flag_outlined,
                  ),
                  MetricCard(
                    label: 'Best Time',
                    value: SwimTime.fromSeconds(bestTime),
                    icon: Icons.timer_outlined,
                  ),
                  MetricCard(
                    label: 'Average Time',
                    value: SwimTime.fromSeconds(avgTime),
                    icon: Icons.speed_outlined,
                  ),
                ],
              ),
              const SizedBox(height: 24),
              const SectionHeader(title: 'Time Progress'),
              SizedBox(
                height: 220,
                child: Card(
                  child: Padding(
                    padding: const EdgeInsets.all(16),
                    child: _TimeProgressChart(logs: data.raceLogs),
                  ),
                ),
              ),
              const SizedBox(height: 24),
              const SectionHeader(title: 'Recent Sessions'),
              ...data.raceLogs.take(10).map((log) => _SessionTile(log: log)),
            ],
          ),
        );
      },
    );
  }
}

class _SessionTile extends StatelessWidget {
  const _SessionTile({required this.log});

  final dynamic log;

  @override
  Widget build(BuildContext context) {
    return Card(
      margin: const EdgeInsets.only(bottom: 10),
      child: ListTile(
        contentPadding: const EdgeInsets.symmetric(horizontal: 16, vertical: 4),
        leading: CircleAvatar(
          backgroundColor: SwimIQTheme.lightSky,
          child: Text(
            log.course,
            style: const TextStyle(
              fontSize: 11,
              fontWeight: FontWeight.w800,
              color: SwimIQTheme.accentBlue,
            ),
          ),
        ),
        title: Text(
          log.event,
          style: const TextStyle(fontWeight: FontWeight.w700),
        ),
        subtitle: Text('${log.date} · ${log.stroke}'),
        trailing: Text(
          SwimTime.fromSeconds(log.timeSeconds),
          style: Theme.of(context).textTheme.titleMedium?.copyWith(
                fontWeight: FontWeight.w900,
                color: SwimIQTheme.darkNavy,
              ),
        ),
      ),
    );
  }
}

class _TimeProgressChart extends StatelessWidget {
  const _TimeProgressChart({required this.logs});

  final List<dynamic> logs;

  @override
  Widget build(BuildContext context) {
    final strokeColors = {
      'Freestyle': SwimIQTheme.primaryBlue,
      'Backstroke': SwimIQTheme.accentBlue,
      'Breaststroke': const Color(0xFF00B894),
      'Butterfly': const Color(0xFFFF7675),
      'IM': const Color(0xFF6C5CE7),
      'Free': SwimIQTheme.primaryBlue,
    };

    final sorted = [...logs]
      ..sort((a, b) => a.date.compareTo(b.date));

    final strokes = sorted.map((l) => l.stroke.toString()).toSet().toList();
    final lineBars = <LineChartBarData>[];

    for (var i = 0; i < strokes.length; i++) {
      final stroke = strokes[i];
      final strokeLogs = sorted.where((l) => l.stroke == stroke).toList();
      final spots = <FlSpot>[];

      for (var j = 0; j < strokeLogs.length; j++) {
        spots.add(FlSpot(j.toDouble(), strokeLogs[j].timeSeconds));
      }

      if (spots.isNotEmpty) {
        lineBars.add(
          LineChartBarData(
            spots: spots,
            isCurved: true,
            color: strokeColors[stroke] ?? SwimIQTheme.primaryBlue,
            barWidth: 3,
            dotData: const FlDotData(show: true),
            belowBarData: BarAreaData(show: false),
          ),
        );
      }
    }

    if (lineBars.isEmpty) {
      return const Center(child: Text('Not enough data for chart'));
    }

    final allTimes = sorted.map((l) => l.timeSeconds).toList();
    final minY = allTimes.reduce((a, b) => a < b ? a : b) - 5;
    final maxY = allTimes.reduce((a, b) => a > b ? a : b) + 5;

    return LineChart(
      LineChartData(
        minY: minY > 0 ? minY : 0,
        maxY: maxY,
        gridData: FlGridData(
          show: true,
          drawVerticalLine: false,
          horizontalInterval: 10,
          getDrawingHorizontalLine: (value) => FlLine(
            color: Colors.grey.shade200,
            strokeWidth: 1,
          ),
        ),
        titlesData: FlTitlesData(
          leftTitles: AxisTitles(
            sideTitles: SideTitles(
              showTitles: true,
              reservedSize: 42,
              getTitlesWidget: (value, meta) {
                return Text(
                  SwimTime.fromSeconds(value),
                  style: TextStyle(fontSize: 10, color: Colors.grey.shade600),
                );
              },
            ),
          ),
          bottomTitles: const AxisTitles(
            sideTitles: SideTitles(showTitles: false),
          ),
          topTitles: const AxisTitles(
            sideTitles: SideTitles(showTitles: false),
          ),
          rightTitles: const AxisTitles(
            sideTitles: SideTitles(showTitles: false),
          ),
        ),
        borderData: FlBorderData(show: false),
        lineBarsData: lineBars,
      ),
    );
  }
}

File: android/app/src/main/java/com/swimiq/app/ui/screens/dashboard/DashboardScreen.kt
package com.swimiq.app.ui.screens.dashboard

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.lazy.LazyColumn
import androidx.compose.foundation.lazy.items
import androidx.compose.material3.Card
import androidx.compose.material3.CardDefaults
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.swimiq.app.ui.components.MetricCard
import com.swimiq.app.ui.components.SectionHeader
import com.swimiq.app.ui.theme.SwimBlueDark
import com.swimiq.app.ui.theme.SwimNavy
import com.swimiq.app.ui.viewmodel.SwimUiState
import com.swimiq.app.util.SwimAnalytics
import com.swimiq.app.util.SwimTimeUtils

@Composable
fun DashboardScreen(
    state: SwimUiState,
    contentPadding: PaddingValues,
) {
    val logs = state.raceLogs
    val breakdown = state.scoreBreakdown

    LazyColumn(
        modifier = Modifier.fillMaxSize(),
        contentPadding = contentPadding,
        verticalArrangement = Arrangement.spacedBy(12.dp),
    ) {
        item {
            Column(modifier = Modifier.padding(horizontal = 16.dp)) {
                Text(
                    text = "Welcome, ${state.displayName}",
                    style = MaterialTheme.typography.headlineSmall,
                    fontWeight = FontWeight.Bold,
                    color = SwimNavy,
                )
                Text(
                    text = "SwimIQ Dashboard · Version 3",
                    style = MaterialTheme.typography.bodyMedium,
                    color = SwimNavy.copy(alpha = 0.7f),
                )
                if (state.isFromCache) {
                    Text(
                        text = "Showing cached data",
                        style = MaterialTheme.typography.labelSmall,
                        color = SwimBlueDark,
                        modifier = Modifier.padding(top = 4.dp),
                    )
                }
            }
        }

        if (logs.isEmpty()) {
            item {
                Text(
                    text = "No swim sessions yet. Add a swim session to start building the dashboard.",
                    modifier = Modifier.padding(horizontal = 16.dp),
                    color = SwimNavy.copy(alpha = 0.7f),
                )
            }
        } else {
            item {
                Column(
                    modifier = Modifier.padding(horizontal = 16.dp),
                    verticalArrangement = Arrangement.spacedBy(8.dp),
                ) {
                    MetricCard("SwimIQ Score™", breakdown.totalScore.toString())
                    MetricCard("Total Sessions", logs.size.toString())
                    MetricCard("Personal Bests", state.personalBestCount.toString())
                    MetricCard("Active Goals", state.goals.size.toString())
                    MetricCard("Best Time", SwimAnalytics.bestTime(logs))
                    MetricCard("Average Time", SwimAnalytics.averageTime(logs))
                }
            }

            item {
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.primaryContainer,
                    ),
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "How your SwimIQ Score works",
                            fontWeight = FontWeight.Bold,
                            color = SwimNavy,
                        )
                        breakdown.explanation.forEach { line ->
                            Text(
                                text = line,
                                color = SwimNavy.copy(alpha = 0.8f),
                                modifier = Modifier.padding(top = 4.dp),
                            )
                        }
                    }
                }
            }

            item {
                SectionHeader(
                    title = "Recent Sessions",
                    modifier = Modifier.padding(horizontal = 16.dp),
                )
            }

            items(logs.take(5)) { log ->
                Card(
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(horizontal = 16.dp),
                    colors = CardDefaults.cardColors(
                        containerColor = MaterialTheme.colorScheme.surface,
                    ),
                ) {
                    Column(modifier = Modifier.padding(16.dp)) {
                        Text(
                            text = "${log.distance} ${log.stroke} · ${log.course}",
                            fontWeight = FontWeight.Bold,
                            color = SwimNavy,
                        )
                        Text(
                            text = "${log.date} · ${SwimTimeUtils.formatSeconds(log.timeSeconds)}",
                            color = SwimNavy.copy(alpha = 0.7f),
                        )
                    }
                }
            }
        }
    }
}
